ykcompile/src/lib.rs

```diff
@@ -0,0 +1,161 @@
+#![feature(proc_macro_hygiene)]
+#![feature(test)]
+
+#[macro_use]
+extern crate dynasm;
+extern crate dynasmrt;
+extern crate test;
+
+use std::collections::HashMap;
+use std::mem;
+
+use yktrace::tir::{
+    Constant, ConstantInt, Operand, Rvalue, Statement, TirOp, TirTrace, UnsignedInt,
+};
+
+use dynasmrt::DynasmApi;
+
+/// A compiled SIRTrace.
+pub struct CompiledTrace {
+    /// A compiled trace.
+    mc: dynasmrt::ExecutableBuffer,
+}
+
+impl CompiledTrace {
+    pub fn execute(&self) -> u64 {
+        // For now a compiled trace always returns whatever has been left in register RAX. We also
+        // assume for now that this will be a `u64`.
+        let func: fn() -> u64 = unsafe { mem::transmute(self.mc.ptr(dynasmrt::AssemblyOffset(0))) };
+        func()
+    }
+}
+
+/// The `TraceCompiler` takes a `SIRTrace` and compiles it to machine code. Returns a `CompiledTrace`.
+pub struct TraceCompiler {
+    asm: dynasmrt::x64::Assembler,
+    /// Contains the list of currently available registers.
+    available_regs: Vec<u8>,
+    /// Maps locals to their assigned registers.
+    assigned_regs: HashMap<u32, u8>,
+}
+
+impl TraceCompiler {
+    fn local_to_reg(&mut self, l: u32) -> u8 {
+        // This is a really dumb register allocator, which runs out of available registers after 7
+        // locals. We can do better than this by using StorageLive/StorageDead from the MIR to free
+        // up registers again, and allocate additional locals on the stack. Though, ultimately we
+        // probably want to implement a proper register allocator, e.g. graph colouring or linear
+        // scan.
+        if l == 0 {
+            0
+        } else {
+            let reg = self
+                .available_regs
+                .pop()
+                .expect("Can't allocate more than 7 locals yet!");
+            *self.assigned_regs.entry(l).or_insert(reg)
+        }
+    }
+
+    /// Move constant `c` of type `usize` into local `a`.
+    pub fn local_usize(&mut self, a: u32, c: usize) {
+        let reg = self.local_to_reg(a);
+        dynasm!(self.asm
+            ; mov Rq(reg), c as i32
```

Comment:
Is the cast here safe? `c` is a `u64`.

Should we be using `try_from`?

Reply:
For some reason dynasm tells me it expects an `i32` here, though we should be able to pass it an `i64`. I'm looking into why that is.

Reply:
Fixed in 78cbcdb